In [1]:
import torch
torch.cuda.empty_cache()

In [2]:
import sys, os
sys.path.append(os.path.abspath(".."))

In [3]:
from src.train import train_model
from src.config import MODEL_LIST

for model_name, model_dir in MODEL_LIST.items():
    print("Loading model:", repr(model_name))  #  Add this debug print
    trainer, model, tokenizer = train_model(model_name, model_dir)
    print(f"Finished training {model_name}. Saved to {model_dir}")


Loading model: 'xlm-roberta-base'

Label distribution:
B-PRODUCT: 58
I-PRODUCT: 58
O: 3372
B-LOC: 65
I-LOC: 539
B-PRICE: 30
I-PRICE: 83


Casting the dataset:   0%|          | 0/99 [00:00<?, ? examples/s]

/home/nurye/Desktop/10_Academy/week_4/Amharic-E-commerce-Data-Extractor/AE-venv-py310/lib/python3.10/site-packages/huggingface_hub/file_download.py:943: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Map:   0%|          | 0/79 [00:00<?, ? examples/s]

Map:   0%|          | 0/20 [00:00<?, ? examples/s]

  0%|          | 0/200 [00:00<?, ?it/s]

  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 108.69105529785156, 'eval_precision': 0.12031691438398232, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.13062177029504918, 'eval_runtime': 0.3961, 'eval_samples_per_second': 50.49, 'eval_steps_per_second': 7.573, 'epoch': 1.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 68.08946228027344, 'eval_precision': 0.12031691438398232, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.13062177029504918, 'eval_runtime': 0.4047, 'eval_samples_per_second': 49.425, 'eval_steps_per_second': 7.414, 'epoch': 2.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 40.08336639404297, 'eval_precision': 0.23486407231410958, 'eval_recall': 0.25735674648261714, 'eval_f1': 0.24538599430913707, 'eval_runtime': 0.3921, 'eval_samples_per_second': 51.009, 'eval_steps_per_second': 7.651, 'epoch': 3.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 29.309194564819336, 'eval_precision': 0.38211538926722205, 'eval_recall': 0.28363015845571543, 'eval_f1': 0.2649151782270495, 'eval_runtime': 0.3913, 'eval_samples_per_second': 51.114, 'eval_steps_per_second': 7.667, 'epoch': 4.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 20.517202377319336, 'eval_precision': 0.7105284861407936, 'eval_recall': 0.6340527937795083, 'eval_f1': 0.6529743119884743, 'eval_runtime': 0.4168, 'eval_samples_per_second': 47.988, 'eval_steps_per_second': 7.198, 'epoch': 5.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 21.51284408569336, 'eval_precision': 0.6824375845593599, 'eval_recall': 0.6882887565474709, 'eval_f1': 0.6567478440445875, 'eval_runtime': 0.3752, 'eval_samples_per_second': 53.3, 'eval_steps_per_second': 7.995, 'epoch': 6.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 14.470846176147461, 'eval_precision': 0.7216333680056375, 'eval_recall': 0.7322139408712206, 'eval_f1': 0.7162434852029234, 'eval_runtime': 0.3813, 'eval_samples_per_second': 52.448, 'eval_steps_per_second': 7.867, 'epoch': 7.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 18.385986328125, 'eval_precision': 0.8877043873839853, 'eval_recall': 0.8889494099036792, 'eval_f1': 0.8851182677107102, 'eval_runtime': 0.372, 'eval_samples_per_second': 53.761, 'eval_steps_per_second': 8.064, 'epoch': 8.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 14.285740852355957, 'eval_precision': 0.9083742730906744, 'eval_recall': 0.91414448027486, 'eval_f1': 0.9095455756513403, 'eval_runtime': 0.3835, 'eval_samples_per_second': 52.149, 'eval_steps_per_second': 7.822, 'epoch': 9.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 13.455101013183594, 'eval_precision': 0.9208256898022361, 'eval_recall': 0.9312481602251184, 'eval_f1': 0.9251018670773499, 'eval_runtime': 0.3992, 'eval_samples_per_second': 50.098, 'eval_steps_per_second': 7.515, 'epoch': 10.0}
{'train_runtime': 139.5858, 'train_samples_per_second': 5.66, 'train_steps_per_second': 1.433, 'train_loss': 39.18659423828125, 'epoch': 10.0}
Finished training xlm-roberta-base. Saved to ../models/xlm-roberta-ner/checkpoint-200
Loading model: 'rasyosef/bert-tiny-amharic'

Label distribution:
B-PRODUCT: 58
I-PRODUCT: 58
O: 3372
B-LOC: 65
I-LOC: 539
B-PRICE: 30
I-PRICE: 83


Casting the dataset:   0%|          | 0/99 [00:00<?, ? examples/s]

/home/nurye/Desktop/10_Academy/week_4/Amharic-E-commerce-Data-Extractor/AE-venv-py310/lib/python3.10/site-packages/huggingface_hub/file_download.py:943: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Map:   0%|          | 0/79 [00:00<?, ? examples/s]

Map:   0%|          | 0/20 [00:00<?, ? examples/s]

Some weights of BertModel were not initialized from the model checkpoint at rasyosef/bert-tiny-amharic and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


  0%|          | 0/200 [00:00<?, ?it/s]

  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 135.71328735351562, 'eval_precision': 0.12449731052888868, 'eval_recall': 0.12201525746207381, 'eval_f1': 0.04593624470656171, 'eval_runtime': 0.1883, 'eval_samples_per_second': 106.241, 'eval_steps_per_second': 15.936, 'epoch': 1.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 101.205078125, 'eval_precision': 0.13774583483234495, 'eval_recall': 0.16121687190486889, 'eval_f1': 0.11227815652220892, 'eval_runtime': 0.1318, 'eval_samples_per_second': 151.724, 'eval_steps_per_second': 22.759, 'epoch': 2.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 77.0671615600586, 'eval_precision': 0.17591118519628407, 'eval_recall': 0.14937331081527228, 'eval_f1': 0.12990213117423138, 'eval_runtime': 0.1359, 'eval_samples_per_second': 147.149, 'eval_steps_per_second': 22.072, 'epoch': 3.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 62.11663818359375, 'eval_precision': 0.11597605352740932, 'eval_recall': 0.14213782189613006, 'eval_f1': 0.1277310924369748, 'eval_runtime': 0.136, 'eval_samples_per_second': 147.016, 'eval_steps_per_second': 22.052, 'epoch': 4.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 53.4643440246582, 'eval_precision': 0.1157796451914099, 'eval_recall': 0.1427132786649403, 'eval_f1': 0.12784328887170565, 'eval_runtime': 0.1479, 'eval_samples_per_second': 135.256, 'eval_steps_per_second': 20.288, 'epoch': 5.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 48.604705810546875, 'eval_precision': 0.11580174927113702, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.12791446605693674, 'eval_runtime': 0.1407, 'eval_samples_per_second': 142.147, 'eval_steps_per_second': 21.322, 'epoch': 6.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 45.79034423828125, 'eval_precision': 0.11580174927113702, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.12791446605693674, 'eval_runtime': 0.1366, 'eval_samples_per_second': 146.373, 'eval_steps_per_second': 21.956, 'epoch': 7.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 43.85332489013672, 'eval_precision': 0.11580174927113702, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.12791446605693674, 'eval_runtime': 0.1407, 'eval_samples_per_second': 142.096, 'eval_steps_per_second': 21.314, 'epoch': 8.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 42.799049377441406, 'eval_precision': 0.11580174927113702, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.12791446605693674, 'eval_runtime': 0.146, 'eval_samples_per_second': 137.016, 'eval_steps_per_second': 20.552, 'epoch': 9.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 42.32746505737305, 'eval_precision': 0.11580174927113702, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.12791446605693674, 'eval_runtime': 0.1391, 'eval_samples_per_second': 143.826, 'eval_steps_per_second': 21.574, 'epoch': 10.0}
{'train_runtime': 30.449, 'train_samples_per_second': 25.945, 'train_steps_per_second': 6.568, 'train_loss': 89.02193359375, 'epoch': 10.0}
Finished training rasyosef/bert-tiny-amharic. Saved to ../models/bert-tiny-amharic/checkpoint-200
Loading model: 'Davlan/afro-xlmr-base'

Label distribution:
B-PRODUCT: 58
I-PRODUCT: 58
O: 3372
B-LOC: 65
I-LOC: 539
B-PRICE: 30
I-PRICE: 83


Casting the dataset:   0%|          | 0/99 [00:00<?, ? examples/s]

/home/nurye/Desktop/10_Academy/week_4/Amharic-E-commerce-Data-Extractor/AE-venv-py310/lib/python3.10/site-packages/huggingface_hub/file_download.py:943: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Map:   0%|          | 0/79 [00:00<?, ? examples/s]

Map:   0%|          | 0/20 [00:00<?, ? examples/s]

Some weights of XLMRobertaModel were not initialized from the model checkpoint at Davlan/afro-xlmr-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


  0%|          | 0/200 [00:00<?, ?it/s]

  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 74.1200942993164, 'eval_precision': 0.11197151757899422, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.1255426375929345, 'eval_runtime': 0.3528, 'eval_samples_per_second': 56.683, 'eval_steps_per_second': 8.502, 'epoch': 1.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 48.90926742553711, 'eval_precision': 0.11197151757899422, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.1255426375929345, 'eval_runtime': 0.3932, 'eval_samples_per_second': 50.865, 'eval_steps_per_second': 7.63, 'epoch': 2.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 23.455154418945312, 'eval_precision': 0.25528769565466813, 'eval_recall': 0.27942656713542496, 'eval_f1': 0.2666266570783005, 'eval_runtime': 0.335, 'eval_samples_per_second': 59.694, 'eval_steps_per_second': 8.954, 'epoch': 3.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 17.587406158447266, 'eval_precision': 0.2622019289872365, 'eval_recall': 0.28389734272087214, 'eval_f1': 0.2725399228899719, 'eval_runtime': 0.3618, 'eval_samples_per_second': 55.273, 'eval_steps_per_second': 8.291, 'epoch': 4.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 15.722715377807617, 'eval_precision': 0.405466733788427, 'eval_recall': 0.3612635562738561, 'eval_f1': 0.37330947511413143, 'eval_runtime': 0.3225, 'eval_samples_per_second': 62.023, 'eval_steps_per_second': 9.304, 'epoch': 5.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 13.677871704101562, 'eval_precision': 0.7724122737228685, 'eval_recall': 0.6122632972531553, 'eval_f1': 0.6443447506003764, 'eval_runtime': 0.3221, 'eval_samples_per_second': 62.099, 'eval_steps_per_second': 9.315, 'epoch': 6.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 11.972696304321289, 'eval_precision': 0.7335669298156452, 'eval_recall': 0.7335849191832964, 'eval_f1': 0.7330135372174187, 'eval_runtime': 0.3523, 'eval_samples_per_second': 56.773, 'eval_steps_per_second': 8.516, 'epoch': 7.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 10.858081817626953, 'eval_precision': 0.7654879449250911, 'eval_recall': 0.782836937197992, 'eval_f1': 0.7697726383457483, 'eval_runtime': 0.3407, 'eval_samples_per_second': 58.707, 'eval_steps_per_second': 8.806, 'epoch': 8.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 10.659294128417969, 'eval_precision': 0.7315917076330917, 'eval_recall': 0.7683747497134921, 'eval_f1': 0.7425172468085527, 'eval_runtime': 0.3713, 'eval_samples_per_second': 53.861, 'eval_steps_per_second': 8.079, 'epoch': 9.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 10.618353843688965, 'eval_precision': 0.7488755967985139, 'eval_recall': 0.7778985592373016, 'eval_f1': 0.755301008982495, 'eval_runtime': 0.3218, 'eval_samples_per_second': 62.148, 'eval_steps_per_second': 9.322, 'epoch': 10.0}
{'train_runtime': 140.7546, 'train_samples_per_second': 5.613, 'train_steps_per_second': 1.421, 'train_loss': 39.1750244140625, 'epoch': 10.0}
Finished training Davlan/afro-xlmr-base. Saved to ../models/afroxlmr-ner/checkpoint-200
Loading model: 'google-bert/bert-base-multilingual-cased'

Label distribution:
B-PRODUCT: 58
I-PRODUCT: 58
O: 3372
B-LOC: 65
I-LOC: 539
B-PRICE: 30
I-PRICE: 83


Casting the dataset:   0%|          | 0/99 [00:00<?, ? examples/s]

/home/nurye/Desktop/10_Academy/week_4/Amharic-E-commerce-Data-Extractor/AE-venv-py310/lib/python3.10/site-packages/huggingface_hub/file_download.py:943: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Map:   0%|          | 0/79 [00:00<?, ? examples/s]

Map:   0%|          | 0/20 [00:00<?, ? examples/s]

  0%|          | 0/200 [00:00<?, ?it/s]

  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 30.56558609008789, 'eval_precision': 0.11508398469981707, 'eval_recall': 0.14285714285714285, 'eval_f1': 0.1274753615179147, 'eval_runtime': 0.1863, 'eval_samples_per_second': 107.36, 'eval_steps_per_second': 16.104, 'epoch': 1.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 25.76980972290039, 'eval_precision': 0.2302089407191448, 'eval_recall': 0.21203904151881028, 'eval_f1': 0.21726424637518266, 'eval_runtime': 0.1872, 'eval_samples_per_second': 106.82, 'eval_steps_per_second': 16.023, 'epoch': 2.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 22.149085998535156, 'eval_precision': 0.24962875573007462, 'eval_recall': 0.2115294485236682, 'eval_f1': 0.22162812994715603, 'eval_runtime': 0.1899, 'eval_samples_per_second': 105.324, 'eval_steps_per_second': 15.799, 'epoch': 3.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 16.26694107055664, 'eval_precision': 0.29574039603037117, 'eval_recall': 0.39303196673138874, 'eval_f1': 0.32996325900431145, 'eval_runtime': 0.2013, 'eval_samples_per_second': 99.368, 'eval_steps_per_second': 14.905, 'epoch': 4.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 14.77794075012207, 'eval_precision': 0.3121071761507404, 'eval_recall': 0.3997831440027972, 'eval_f1': 0.34373121965041264, 'eval_runtime': 0.1831, 'eval_samples_per_second': 109.227, 'eval_steps_per_second': 16.384, 'epoch': 5.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 12.797861099243164, 'eval_precision': 0.387507678656747, 'eval_recall': 0.3984477871761109, 'eval_f1': 0.3905868982683854, 'eval_runtime': 0.2348, 'eval_samples_per_second': 85.185, 'eval_steps_per_second': 12.778, 'epoch': 6.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 14.636128425598145, 'eval_precision': 0.5197639430823288, 'eval_recall': 0.4586355840691101, 'eval_f1': 0.4069839864514952, 'eval_runtime': 0.1885, 'eval_samples_per_second': 106.091, 'eval_steps_per_second': 15.914, 'epoch': 7.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 11.8098726272583, 'eval_precision': 0.7522411569548224, 'eval_recall': 0.47535023430977186, 'eval_f1': 0.4526444605454682, 'eval_runtime': 0.1831, 'eval_samples_per_second': 109.239, 'eval_steps_per_second': 16.386, 'epoch': 8.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 10.895940780639648, 'eval_precision': 0.7569989534275249, 'eval_recall': 0.49410142531529816, 'eval_f1': 0.4861420584150848, 'eval_runtime': 0.1877, 'eval_samples_per_second': 106.542, 'eval_steps_per_second': 15.981, 'epoch': 9.0}


  0%|          | 0/3 [00:00<?, ?it/s]

{'eval_loss': 11.097259521484375, 'eval_precision': 0.6348895357516048, 'eval_recall': 0.5333871396010125, 'eval_f1': 0.528353257131412, 'eval_runtime': 0.1886, 'eval_samples_per_second': 106.061, 'eval_steps_per_second': 15.909, 'epoch': 10.0}
{'train_runtime': 80.6455, 'train_samples_per_second': 9.796, 'train_steps_per_second': 2.48, 'train_loss': 24.244951171875, 'epoch': 10.0}
Finished training google-bert/bert-base-multilingual-cased. Saved to ../models/mbert-ner/checkpoint-200


In [9]:
from transformers import AutoConfig
import os

# Define model directories and their base models
MODELS = {
    "xlm-roberta": {
        "path": "../models/xlm-roberta-ner/checkpoint-200",
        "base": "xlm-roberta-base"
    },
    "bert-tiny-amharic": {
        "path": "../models/bert-tiny-amharic/checkpoint-200",
        "base": "rasyosef/bert-tiny-amharic"
    },
    "afroxlmr": {
        "path": "../models/afroxlmr-ner/checkpoint-200",
        "base": "Davlan/afro-xlmr-base"
    },
    "mbert": {
        "path": "../models/mbert-ner/checkpoint-200",
        "base": "bert-base-multilingual-cased"
    }
}

def ensure_config(model_name, model_info):
    config_path = os.path.join(model_info["path"], "config.json")
    if os.path.exists(config_path):
        print(f" {model_name}: config.json already exists.")
    else:
        print(f" {model_name}: config.json missing. Generating from base model {model_info['base']}...")
        config = AutoConfig.from_pretrained(model_info["base"])
        config.save_pretrained(model_info["path"])
        print(f" {model_name}: config.json generated.")

if __name__ == "__main__":
    for name, info in MODELS.items():
        ensure_config(name, info)


 xlm-roberta: config.json missing. Generating from base model xlm-roberta-base...


/home/nurye/Desktop/10_Academy/week_4/Amharic-E-commerce-Data-Extractor/AE-venv-py310/lib/python3.10/site-packages/huggingface_hub/file_download.py:943: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


 xlm-roberta: config.json generated.
 bert-tiny-amharic: config.json missing. Generating from base model rasyosef/bert-tiny-amharic...
 bert-tiny-amharic: config.json generated.
 afroxlmr: config.json missing. Generating from base model Davlan/afro-xlmr-base...
 afroxlmr: config.json generated.
 mbert: config.json missing. Generating from base model bert-base-multilingual-cased...


config.json:   0%|          | 0.00/625 [00:00<?, ?B/s]

 mbert: config.json generated.


In [16]:
from transformers import AutoTokenizer

MODEL_BASES = {
    "xlm-roberta-ner": "xlm-roberta-base",
    "bert-tiny-amharic": "rasyosef/bert-tiny-amharic",
    "afroxlmr-ner": "Davlan/afro-xlmr-base",
    "mbert-ner": "bert-base-multilingual-cased",
}

for model_name, base_model_name in MODEL_BASES.items():
    checkpoint_path = f"../models/{model_name}/checkpoint-200"  # adjust path if needed
    print(f"Processing {model_name} - loading tokenizer from {base_model_name}")
    
    tokenizer = AutoTokenizer.from_pretrained(base_model_name)
    tokenizer.save_pretrained(checkpoint_path)
    print(f"Saved tokenizer to {checkpoint_path}\n")


Processing xlm-roberta-ner - loading tokenizer from xlm-roberta-base


/home/nurye/Desktop/10_Academy/week_4/Amharic-E-commerce-Data-Extractor/AE-venv-py310/lib/python3.10/site-packages/huggingface_hub/file_download.py:943: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


Saved tokenizer to ../models/xlm-roberta-ner/checkpoint-200

Processing bert-tiny-amharic - loading tokenizer from rasyosef/bert-tiny-amharic
Saved tokenizer to ../models/bert-tiny-amharic/checkpoint-200

Processing afroxlmr-ner - loading tokenizer from Davlan/afro-xlmr-base
Saved tokenizer to ../models/afroxlmr-ner/checkpoint-200

Processing mbert-ner - loading tokenizer from bert-base-multilingual-cased


tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/996k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

Saved tokenizer to ../models/mbert-ner/checkpoint-200



In [17]:
from src.predict import load_model, predict_text
from src.config import MODEL_LIST
text = "ስልኩ 0911223344 ነው አድራሻችን አዲስ አበባ ህንፃ 12 ዋጋ 45 ብር"
for model_name, path in MODEL_LIST.items():
    print(f"\n Testing: {model_name}")
    model, tokenizer = load_model(model_name, path)
    predictions = predict_text(text, model, tokenizer)
    for word, tag in predictions:
        print(f"{word:<10} → {tag}")



 Testing: xlm-roberta-base
ስልኩ        → B-PRODUCT
0911223344 → O
ነው         → O
አድራሻችን     → B-LOC
አዲስ        → I-LOC
አበባ        → I-LOC
ህንፃ        → I-LOC
12         → I-LOC
ዋጋ         → B-PRICE
45         → I-PRICE
ብር         → I-PRICE

 Testing: rasyosef/bert-tiny-amharic


Some weights of BertModel were not initialized from the model checkpoint at rasyosef/bert-tiny-amharic and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


ስልኩ        → O
0911223344 → O
ነው         → O
አድራሻችን     → B-LOC
አዲስ        → O
አበባ        → O
ህንፃ        → O
12         → O
ዋጋ         → O
45         → O
ብር         → O

 Testing: Davlan/afro-xlmr-base


Some weights of XLMRobertaModel were not initialized from the model checkpoint at Davlan/afro-xlmr-base and are newly initialized: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


ስልኩ        → B-PRODUCT
0911223344 → I-PRODUCT
ነው         → O
አድራሻችን     → O
አዲስ        → O
አበባ        → O
ህንፃ        → O
12         → O
ዋጋ         → O
45         → O
ብር         → O

 Testing: google-bert/bert-base-multilingual-cased
ስልኩ        → O
0911223344 → O
ነው         → O
አድራሻችን     → B-LOC
አዲስ        → I-LOC
አበባ        → I-LOC
ህንፃ        → O
12         → O
ዋጋ         → O
45         → O
ብር         → O


# #model comparion best model is xlm-roberta-base

| Model Name                          | Eval F1   | Precision   | Recall    | Eval Loss   | Train Loss   | Train Time (s) | Samples/sec | Notes                  |
| ----------------------------------- | --------- | ----------- | --------- | ----------- | ------------ | -------------- | ----------- | ---------------------- |
| **1. xlm-roberta-base**             | **0.925** | **0.921**   | **0.931** | 13.46       | 39.19        | 139.59         | 5.66        |  Highest performance  |
| **2. bert-tiny-amharic**            | 0.128     | 0.116       | 0.143     | 42.33       | 89.02        | 30.45          | **25.95**   |  Weak NER capability |
| **3. afro-xlmr-base**               | 0.755     | 0.749       | 0.778     | **10.62**   | 39.18        | 140.75         | 5.61        |  Good balance         |
| **4. bert-base-multilingual-cased** | 0.528     | 0.635       | 0.533     | 11.10       | 24.24        | 80.65          | 9.80        |  Moderate accuracy   |



## why it is best
Why it's best:
Highest F1 score (0.925): Best balance of precision and recall.
Strong generalization: Despite slightly higher eval loss than AfroXLM-R, F1 score is significantly better.
Good training stability: Train loss is on par with afro-xlmr and lower than underperforming models.
Well-suited for multilingual NER, especially with morphologically rich languages like Amharic."""
## Additional Observations:
afro-xlmr-base is a solid second choice if looking for a model with slightly lower eval loss and smaller model size than XLM-R.
bert-tiny-amharic performed poorly, likely due to small size and limited capacity—not recommended for NER.
mBERT had moderate performance but lower F1 than AfroXLMR and XLM-R, making it a fallback option only."""
## Recommendation:
Use xlm-roberta-base for deployment or further fine-tuning. It provides the best balance between accuracy and generalization for your Amharic NER task.
If you want a slightly faster or smaller model for low-resource deployment, consider afro-xlmr-base as a good trade-off.
Would you like a visualization or exportable table for documentation?"""

In [ ]:
import os
from transformers import pipeline, AutoTokenizer, AutoModelForTokenClassification
from lime.lime_text import LimeTextExplainer
import shap
import torch
import numpy as np

# Set your models and labels
MODEL_PATHS = {
    "xlm-roberta": "../models/xlm-roberta-ner/checkpoint-200/checkpoint-200",
    "bert-tiny-amharic": "../models/bert-tiny-amharic/checkpoint-200/checkpoint-200",
    "afroxlmr": "../models/afroxlmr-ner/checkpoint-200/checkpoint-200",
    "mbert": "../models/mbert-ner/checkpoint-200/checkpoint-200"
}

SAMPLE_TEXT = "አዲስ አበባ ላይ ምርት ተመን 150 ብር ነበር።"  # Amharic sentence

def interpret_with_lime(model_name, pipeline_func):
    print(f"\n🔍 LIME Interpretation for {model_name}")
    explainer = LimeTextExplainer(class_names=["O", "ENTITY"])

    def predict_proba(texts):
        results = []
        for text in texts:
            preds = pipeline_func(text)
            # count how many words were tagged as entities
            entity_count = sum(1 for p in preds if p.get("entity_group", "O") != "O")
            total_words = len(text.split())
            avg = entity_count / total_words if total_words > 0 else 0
            results.append([1 - avg, avg])
        return np.array(results)

    exp = explainer.explain_instance(SAMPLE_TEXT, predict_proba, num_features=10)
    exp.save_to_file(f"{model_name}_lime.html")
def process_model(model_name, model_path):
    print(f"\n🚀 Processing: {model_name}")
    model = AutoModelForTokenClassification.from_pretrained(model_path)
    tokenizer = AutoTokenizer.from_pretrained(model_path)
    ner_pipe = pipeline("ner", model=model, tokenizer=tokenizer, aggregation_strategy="simple")

    # LIME
    interpret_with_lime(model_name, ner_pipe)

# Run for all models
for name, path in MODEL_PATHS.items():
    process_model(name, path)



🚀 Processing: xlm-roberta


Some weights of XLMRobertaForTokenClassification were not initialized from the model checkpoint at ../models/xlm-roberta-ner/checkpoint-200 and are newly initialized: ['classifier.bias', 'classifier.weight', 'embeddings.LayerNorm.bias', 'embeddings.LayerNorm.weight', 'embeddings.position_embeddings.weight', 'embeddings.token_type_embeddings.weight', 'embeddings.word_embeddings.weight', 'encoder.layer.0.attention.output.LayerNorm.bias', 'encoder.layer.0.attention.output.LayerNorm.weight', 'encoder.layer.0.attention.output.dense.bias', 'encoder.layer.0.attention.output.dense.weight', 'encoder.layer.0.attention.self.key.bias', 'encoder.layer.0.attention.self.key.weight', 'encoder.layer.0.attention.self.query.bias', 'encoder.layer.0.attention.self.query.weight', 'encoder.layer.0.attention.self.value.bias', 'encoder.layer.0.attention.self.value.weight', 'encoder.layer.0.intermediate.dense.bias', 'encoder.layer.0.intermediate.dense.weight', 'encoder.layer.0.output.LayerNorm.bias', 'encoder.l


🔍 LIME Interpretation for xlm-roberta


Some weights of BertForTokenClassification were not initialized from the model checkpoint at ../models/bert-tiny-amharic/checkpoint-200 and are newly initialized: ['classifier.bias', 'classifier.weight', 'embeddings.LayerNorm.bias', 'embeddings.LayerNorm.weight', 'embeddings.position_embeddings.weight', 'embeddings.token_type_embeddings.weight', 'embeddings.word_embeddings.weight', 'encoder.layer.0.attention.output.LayerNorm.bias', 'encoder.layer.0.attention.output.LayerNorm.weight', 'encoder.layer.0.attention.output.dense.bias', 'encoder.layer.0.attention.output.dense.weight', 'encoder.layer.0.attention.self.key.bias', 'encoder.layer.0.attention.self.key.weight', 'encoder.layer.0.attention.self.query.bias', 'encoder.layer.0.attention.self.query.weight', 'encoder.layer.0.attention.self.value.bias', 'encoder.layer.0.attention.self.value.weight', 'encoder.layer.0.intermediate.dense.bias', 'encoder.layer.0.intermediate.dense.weight', 'encoder.layer.0.output.LayerNorm.bias', 'encoder.layer


🚀 Processing: bert-tiny-amharic

🔍 LIME Interpretation for bert-tiny-amharic

🚀 Processing: afroxlmr


Some weights of XLMRobertaForTokenClassification were not initialized from the model checkpoint at ../models/afroxlmr-ner/checkpoint-200 and are newly initialized: ['classifier.bias', 'classifier.weight', 'embeddings.LayerNorm.bias', 'embeddings.LayerNorm.weight', 'embeddings.position_embeddings.weight', 'embeddings.token_type_embeddings.weight', 'embeddings.word_embeddings.weight', 'encoder.layer.0.attention.output.LayerNorm.bias', 'encoder.layer.0.attention.output.LayerNorm.weight', 'encoder.layer.0.attention.output.dense.bias', 'encoder.layer.0.attention.output.dense.weight', 'encoder.layer.0.attention.self.key.bias', 'encoder.layer.0.attention.self.key.weight', 'encoder.layer.0.attention.self.query.bias', 'encoder.layer.0.attention.self.query.weight', 'encoder.layer.0.attention.self.value.bias', 'encoder.layer.0.attention.self.value.weight', 'encoder.layer.0.intermediate.dense.bias', 'encoder.layer.0.intermediate.dense.weight', 'encoder.layer.0.output.LayerNorm.bias', 'encoder.laye


🔍 LIME Interpretation for afroxlmr

🚀 Processing: mbert


Some weights of BertForTokenClassification were not initialized from the model checkpoint at ../models/mbert-ner/checkpoint-200 and are newly initialized: ['classifier.bias', 'classifier.weight', 'embeddings.LayerNorm.bias', 'embeddings.LayerNorm.weight', 'embeddings.position_embeddings.weight', 'embeddings.token_type_embeddings.weight', 'embeddings.word_embeddings.weight', 'encoder.layer.0.attention.output.LayerNorm.bias', 'encoder.layer.0.attention.output.LayerNorm.weight', 'encoder.layer.0.attention.output.dense.bias', 'encoder.layer.0.attention.output.dense.weight', 'encoder.layer.0.attention.self.key.bias', 'encoder.layer.0.attention.self.key.weight', 'encoder.layer.0.attention.self.query.bias', 'encoder.layer.0.attention.self.query.weight', 'encoder.layer.0.attention.self.value.bias', 'encoder.layer.0.attention.self.value.weight', 'encoder.layer.0.intermediate.dense.bias', 'encoder.layer.0.intermediate.dense.weight', 'encoder.layer.0.output.LayerNorm.bias', 'encoder.layer.0.outpu


🔍 LIME Interpretation for mbert
